core/shared/src/main/scala/zio/FiberRef.scala

```diff
@@ -103,9 +109,48 @@ final class FiberRef[A](private[zio] val initial: A, private[zio] val combine: (
     (result, result)
   }
 
+  /**
+   * Returns a handle that can be used to read the value of this `FiberRef` from side effecting code.
+   *
+   * This feature is meant to be used for integration with side effecting code, that needs to access fiber specific data.
+   */
+  final def getUnsafeHandle: UIO[UnsafeHandle[A]] =
+    get.flatMap { a =>
+      ZIO.effectTotal {
+        val (threadLocal, isNew) = getThreadLocal
+        threadLocal.set(a)
+        new UnsafeHandle[A] {
+          def get(): A = threadLocal.get()
+        } -> isNew
+      }.flatMap {
+        case (handle, isNew) =>
+          if (!isNew) ZIO.succeed(handle)
+          else
+            ZIO.effectTotal {
+              UnsafelyExposedFiberRefs.register(this)
+              handle
+            }
+      }
+    }
+
+  private[this] def getThreadLocal: (ThreadLocal[A], Boolean) = {
+    val candidate = threadLocalRef.get()
+    if (candidate ne null) (candidate, false)
+    else {
+      val newCandidate = new ThreadLocal[A] {
+        override def initialValue(): A = initial
+      }
+
+      if (!threadLocalRef.compareAndSet(null, newCandidate)) getThreadLocal
+      else (newCandidate, true)
+    }
+  }
 }
 
 object FiberRef extends Serializable {
+  trait UnsafeHandle[A] {
```

Comment:
The `get` should be renamed `unsafeGet` by ZIO conventions.

Also I'm not sure we really need it, since we could just return `Function0[A]` from `unsafeHandle: UIO[Function0[A]]`.

Reply:
I've renamed ` get` to `unsafeGet` as proposed. I'm also fine with `UIO[Function0[A]]`, but I'm slightly more inclined to having a separate trait, as at makes things more explicit.